In [1]:
# Standard imports
import numpy as np
import pandas as pd

# For statistical modeling
import scipy.stats as stats
from math import sqrt 

# For data visualization
import seaborn as sns
import matplotlib.pyplot as plt

# To avoid deprecation and other warnings
import warnings
warnings.filterwarnings('ignore')

# For modeling
# to evaluated performance using rmse
from sklearn.metrics import mean_squared_error
# holt's linear trend model
from statsmodels.tsa.api import Holt
import statsmodels.api as sm

from matplotlib.ticker import StrMethodFormatter
from matplotlib.dates import DateFormatter

# working with dates
from datetime import datetime

import os

url = 'https://www.fhwa.dot.gov/ohim/tvtw/archive/arch92u.htm'
df = pd.read_csv(url)

from requests import get
from bs4 import BeautifulSoup

url = 'https://codeup.com/data-science/math-in-data-science/'
headers = {'User-Agent': 'Codeup Data Science'} # Some websites don't accept the pyhon-requests default user-agent
response = get(url, headers=headers)

soup = BeautifulSoup(response.content, 'html.parser')

s = get('https://www.fhwa.dot.gov/ohim/tvtw/archive/arch92u.htm')

soup.prettify()

response = get('https://www.fhwa.dot.gov/ohim/tvtw/archive/arch92u.htm')
soup = BeautifulSoup(response.content, 'html.parser')

soup.prettify()

soup.title.string

soup.get_text()

soup.select("footer")

In [2]:
#traffic = '/Users/everettclark/Downloads/TMS_traffic_counts/TMS_traffic_counts.csv'
#cspath = '/Users/everettclark/Downloads/archive (1)'
path2 = '/Users/everettclark/Downloads/archive (3)'

df = pd.read_csv(traffic)

df

df.isnull().sum()

df.info()

df.siteDescription.value_counts()

df.flowDirection.value_counts()

sns.scatterplot(x=df.laneNumber, y=df.trafficCount, hue=df.flowDirection)

df = pd.read_csv(f'{cspath}/esea_master_dmg_demos.part1.csv')

df

df.columns

df = pd.read_csv(f'{cspath}/map_data.csv')
df

## Age Specific Fertility Rates

In [3]:
asf = pd.read_csv(f'{path2}/age_specific_fertility_rates.csv')
print(asf.shape)
asf.drop(columns='country_code', inplace=True)
asf.head()

(15106, 13)


country_name  year  fertility_rate_15_19  fertility_rate_20_24  \
0     Slovenia  2036                   8.5                  60.2   
1     Slovenia  2022                   7.5                  56.4   
2     Slovenia  2023                   7.6                  56.5   
3     Slovenia  2024                   7.7                  56.8   
4     Slovenia  2025                   7.7                  56.9   

   fertility_rate_25_29  fertility_rate_30_34  fertility_rate_35_39  \
0                 112.0                  86.3                  32.0   
1                 106.1                  75.5                  26.7   
2                 106.6                  76.2                  27.2   
3                 107.0                  77.1                  27.5   
4                 107.4                  78.0                  27.7   

   fertility_rate_40_44  fertility_rate_45_49  total_fertility_rate  \
0                   8.8                   3.7                1.5555   
1                   6.4                   2.3                1.4110   
2                   6.6                   2.4                1.4213   
3                   6.7                   2.5                1.4316   
4                   6.9                   2.6                1.4419   

   gross_reproduction_rate  sex_ratio_at_birth  
0                   0.7529              1.0660  
1                   0.6829              1.0661  
2                   0.6879              1.0660  
3                   0.6929              1.0660  
4                   0.6979              1.0661

#### I have other info that I can use for birth/death

---

## Country Areas:

In [4]:
pd.read_csv(f'{path2}/country_names_area.csv').head(3)

country_code country_name  country_area
0           AF  Afghanistan      652230.0
1           AL      Albania       27398.0
2           AG      Algeria     2381741.0

#### Not needed

---

## Age 5 Midyear Populations:

In [5]:
mp5 = pd.read_csv(f'{path2}/midyear_population_5yr_age_sex.csv')
mp5.drop(columns='country_code', inplace=True)
print(mp5.shape)
mp5.sort_values(by='country_name').head()

(333080, 9)


country_name  year total_flag  starting_age age_group_indicator  \
279318  Afghanistan  2017          A            95                   -   
244100  Afghanistan  2040          A            15                   -   
193531  Afghanistan  2038          A            50                   -   
50334   Afghanistan  2025          A             0                   -   
127363  Afghanistan  1985          A            25                   -   

        ending_age  midyear_population  midyear_population_male  \
279318          99                 273                       99   
244100          19             5727704                  2908688   
193531          54             1873174                   935379   
50334            4             6055219                  3079206   
127363          29              995440                   510940   

        midyear_population_female  
279318                        174  
244100                    2819016  
193531                     937795  
50334                     2976013  
127363                     484500

#### Not going to use

---

## Midyear Population country code:

In [6]:
mpc = pd.read_csv(f'{path2}/midyear_population_age_country_code.csv')
mpc.sort_values(by='country_name').head(3)

country_code country_name  year   sex  max_age  population_age_0  \
2460213           AF  Afghanistan  2050  Male      100            699328   
2201225           AF  Afghanistan  2024  Male      100            653702   
65574             AF  Afghanistan  2014  Male      100            573665   

         population_age_1  population_age_2  population_age_3  \
2460213            685484            678841            673365   
2201225            620418            604670            592017   
65574              528794            505300            487940   

         population_age_4  ...  population_age_94  population_age_95  \
2460213            668619  ...                586                388   
2201225            579188  ...                107                 65   
65574              472797  ...                 70                 43   

         population_age_96  population_age_97  population_age_98  \
2460213                240                143                 82   
2201225                 35                 21                 11   
65574                   24                 10                  6   

         population_age_99  population_age_100  age  permutation_role  \
2460213                 49                  59   88             child   
2201225                  7                   6   61             child   
65574                    4                   3   29             child   

         population  
2460213        4827  
2201225       73743  
65574        220315  

[3 rows x 109 columns]

mpc.sort_values(by='country_name').head(15)

mpc.country_name.value_counts()

mpc[(mpc.country_name =='Afghanistan') & (mpc.sex == 'Male') & (mpc.year == 1979)].describe()

### Not using

---

## Midyear Population age sex:

In [7]:
mpa = pd.read_csv(f'{path2}/midyear_population_age_sex.csv')
mpa.drop(columns='country_code', inplace=True)
print(mpa.shape)
mpa.head(10)

(30280, 105)


country_name  year     sex  max_age  population_age_0  population_age_1  \
0  Afghanistan  1979    Male      100            318425            299061   
1  Afghanistan  1980  Female      100            303582            253596   
2  Afghanistan  1981    Male      100            278483            253006   
3  Afghanistan  1982    Male      100            260493            226577   
4  Afghanistan  1986    Male      100            288288            244977   
5  Afghanistan  1988    Male      100            305181            262454   
6  Afghanistan  1990    Male      100            324569            283502   
7  Afghanistan  1992  Female      100            354761            310868   
8  Afghanistan  1995  Female      100            447128            395670   
9  Afghanistan  1996  Female      100            463943            411579   

   population_age_2  population_age_3  population_age_4  population_age_5  \
0            282850            268400            255599            244330   
1            255259            249053            239924            232073   
2            228441            233548            226679            220238   
3            217475            198617            205456            200768   
4            225403            207463            204496            195268   
5            238563            218362            205439            190337   
6            259125            240125            223228            206840   
7            285194            268144            251210            232928   
8            363178            348926            332989            316915   
9            381929            356762            344747            330183   

   ...  population_age_91  population_age_92  population_age_93  \
0  ...                170                109                 67   
1  ...                 71                 45                 27   
2  ...                169                108                 67   
3  ...                168                108                 67   
4  ...                159                108                 67   
5  ...                 77                 82                 64   
6  ...                 92                 52                 31   
7  ...                 96                 64                 35   
8  ...                101                 59                 36   
9  ...                118                 69                 38   

   population_age_94  population_age_95  population_age_96  population_age_97  \
0                 42                 24                 14                  8   
1                 16                  9                  6                  3   
2                 41                 24                 14                  8   
3                 41                 24                 14                  7   
4                 40                 25                 13                  8   
5                 42                 23                 14                  9   
6                 32                 24                 14                  7   
7                 18                 12                  7                  6   
8                 26                 17                  8                  4   
9                 23                 16                 10                  5   

   population_age_98  population_age_99  population_age_100  
0                  4                  2                   2  
1                  1                  1                   1  
2                  4                  2                   2  
3                  5                  2                   2  
4                  4                  2                   2  
5                  3                  3                   1  
6                  5                  3                   1  
7                  3                  1                   2  
8                  3                  0                   3  
9                  2                  2                 

In [8]:
mpa[mpa.country_name == 'Afghanistan'].year.value_counts()

1979    2
1980    2
2012    2
2009    2
2005    2
       ..
2021    2
2019    2
2017    2
2016    2
2043    2
Name: year, Length: 72, dtype: int64

In [9]:
mpa[mpa.country_name == 'Afghanistan'].sort_values(by='year')

country_name  year     sex  max_age  population_age_0  population_age_1  \
0      Afghanistan  1979    Male      100            318425            299061   
20162  Afghanistan  1979  Female      100            299276            284224   
1      Afghanistan  1980  Female      100            303582            253596   
20163  Afghanistan  1980    Male      100            313513            270554   
10055  Afghanistan  1981  Female      100            269398            245322   
...            ...   ...     ...      ...               ...               ...   
20203  Afghanistan  2048    Male      100            697565            680954   
10106  Afghanistan  2049  Female      100            670697            657731   
10105  Afghanistan  2049    Male      100            698911            683655   
10107  Afghanistan  2050    Male      100            699328            685484   
48     Afghanistan  2050  Female      100            671003            659370   

       population_age_2  population_age_3  population_age_4  population_age_5  \
0                282850            268400            255599            244330   
20162            270908            258667            247496            237362   
1                255259            249053            239924            232073   
20163            270040            260662            249193            239909   
10055            212463            219832            216190            211718   
...                 ...               ...               ...               ...   
20203            673225            667942            662679            656964   
10106            650804            645829            641695            637315   
10105            675968            670535            665963            661149   
10107            678841            673365            668619            664476   
48               653446            648441            644139            640421   

       ...  population_age_91  population_age_92  population_age_93  \
0      ...                170                109                 67   
20162  ...                 69                 43                 26   
1      ...                 71                 45                 27   
20163  ...                169                108                 68   
10055  ...                 74                 46                 29   
...    ...                ...                ...                ...   
20203  ...               1621               1149                797   
10106  ...               3636               2599               1906   
10105  ...               1764               1195                828   
10107  ...               1918               1303                864   
48     ...               3977               2841               1994   

       population_age_94  population_age_95  population_age_96  \
0                     42                 24                 14   
20162                 15                  9                  5   
1                     16                  9                  6   
20163                 41                 24                 14   
10055                 17                  9                  5   
...                  ...                ...                ...   
20203                515                322                196   
10106               1368                921                599   
10105                562                355                217   
10107                586                388                240   
48                  1434               1008                665   

       population_age_97  population_age_98  population_age_99  \
0                      8                  4                  2   
20162                  3                  2                  0   
1                      3                  1                  1   
20163                  8                  4                  2   
10055                  4                  2                  0   
...               

In [10]:
mpa.columns

Index(['country_name', 'year', 'sex', 'max_age', 'population_age_0',
       'population_age_1', 'population_age_2', 'population_age_3',
       'population_age_4', 'population_age_5',
       ...
       'population_age_91', 'population_age_92', 'population_age_93',
       'population_age_94', 'population_age_95', 'population_age_96',
       'population_age_97', 'population_age_98', 'population_age_99',
       'population_age_100'],
      dtype='object', length=105)

In [11]:
mpa.max_age.min()

80

In [12]:
mpa.iloc[:1]

country_name  year   sex  max_age  population_age_0  population_age_1  \
0  Afghanistan  1979  Male      100            318425            299061   

   population_age_2  population_age_3  population_age_4  population_age_5  \
0            282850            268400            255599            244330   

   ...  population_age_91  population_age_92  population_age_93  \
0  ...                170                109                 67   

   population_age_94  population_age_95  population_age_96  population_age_97  \
0                 42                 24                 14                  8   

   population_age_98  population_age_99  population_age_100  
0                  4                  2                   2  

[1 rows x 105 columns]

mpa.iloc[:1].columns[5:]

---

## Midyear Population:

In [13]:
mp1 = pd.read_csv(f'{path2}/midyear_population.csv')
mp1.drop(columns='country_code', inplace=True)
print(mp1.shape)
mp1.head(10)

(23028, 3)


country_name  year  midyear_population
0     Slovenia  2013             1992690
1     Slovenia  2025             1907560
2     Slovenia  2024             1917093
3     Slovenia  2023             1926279
4     Slovenia  2022             1935080
5     Slovenia  2021             1943453
6     Slovenia  2020             1951358
7     Slovenia  2019             1958778
8     Slovenia  2018             1965706
9     Slovenia  2017             1972126

mp1[(mp1.country_name == 'Norway') &(mp1.year == 2036)]

In [14]:
for col in mp1:
    print(mp1[col].value_counts())
    print('--------')

Slovenia               101
Bangladesh             101
Ireland                101
Virgin Islands U.S.    101
Gambia The             101
                      ... 
Montserrat             101
Panama                 101
Macau                  101
Nepal                  101
Solomon Islands        101
Name: country_name, Length: 228, dtype: int64
--------
2013    228
1999    228
1952    228
1953    228
1954    228
       ... 
2046    228
2047    228
2048    228
2049    228
1975    228
Name: year, Length: 101, dtype: int64
--------
8500        4
8300        4
107580      3
7852        3
7791        3
           ..
27087450    1
27087181    1
27082378    1
27073044    1
193000      1
Name: midyear_population, Length: 22880, dtype: int64
--------


pd.pivot_table(data, values=None, index=None, columns=None, aggfunc='mean', fill_value=None, margins=False, dropna=True, margins_name='All', observed=False, sort=True)

In [50]:
pd.pivot_table(mp1, values='midyear_population', index='country_name', columns='year')

year               1950     1951     1952     1953     1954     1955  \
country_name                                                           
Afghanistan     8150368  8284473  8425333  8573217  8728408  8891209   
Albania         1227156  1254119  1282697  1314608  1352982  1392164   
Algeria         8892718  9073304  9279525  9531710  9611093  9841851   
American Samoa    19100    19500    19300    19200    19500    19800   
Andorra            6176     6310     5866     5591     5503     6189   
...                 ...      ...      ...      ...      ...      ...   
West Bank        771165   774512   777872   781247   784637   788042   
Western Sahara     9057    10133    11336    12682    14187    15872   
Yemen           4777089  4869125  4963829  5061153  5162494  5265327   
Zambia          2553000  2611000  2672000  2734000  2800000  2869000   
Zimbabwe        2853151  2950941  3080907  3190967  3307273  3409017   

year                1956      1957      1958      1959  ...      2041  \
country_name                                            ...             
Afghanistan      9061938   9240934   9428556   9624606  ...  55619723   
Albania          1434476   1476505   1521417   1571329  ...   2978173   
Algeria         10057133  10270856  10484925  10696396  ...  52413932   
American Samoa     20100     20300     20500     20400  ...     49994   
Andorra             6221      6424      7103      7239  ...     80614   
...                  ...       ...       ...       ...  ...       ...   
West Bank         791462    794895    798345    801809  ...   3878368   
Western Sahara     17756     19864     22222     24860  ...   1014876   
Yemen            5380106   5498090   5619377   5743926  ...  41667902   
Zambia           2941000   3016000   3094000   3173000  ...  31141999   
Zimbabwe         3529739   3646340   3764157   3886947  ...  22162763   

year                2042      2043      2044      2045      2046      2047  \
country_name                                                                 
Afghanistan     56524253  57432905  58343854  59255165  60165568  61075545   
Albania          2962049   2945632   2928959   2912050   2894917   2877563   
Algeria         52795174  53170494  53537501  53893778  54237145  54565705   
American Samoa     49827     49683     49561     49461     49384     49332   
Andorra            80099     79549     78963     78341     77686     76998   
...                  ...       ...       ...       ...       ...       ...   
West Bank        3918203   3957435   3996023   4033925   4071177   4107818   
Western Sahara   1032578   1050278   1067959   1085605   1103223   1120822   
Yemen           42188741  42703176  43210273  43709073  44199728  44682648   
Zambia          31959667  32791505  33637166  34496266  35368942  36255336   
Zimbabwe        22490092  22820864  23154725  23491259  23829993  24170434   

year                2048      2049      2050  
country_name                                  
Afghanistan     61985148  62892446  63795418  
Albania          2859983   2842146   2824012  
Algeria         54877585  55171092  55444735  
American Samoa     49303     49296     49308  
Andorra            76280     75534     74765  
...                  ...       ...       ...  
West Bank        4143794   4179053   4213540  
Western Sahara   1138386   1155900   1173350  
Yemen           45157387  45623514  46080625  
Zambia          37155040  38067622  38992619  
Zimbabwe        24512164  24854858  25198196  

[228 rows x 101 columns]

#### Not going to use

---

## Birth Death Growth Rates

In [15]:
bd = pd.read_csv(f'{path2}/birth_death_growth_rates.csv')
bd.drop(columns='country_code', inplace=True)
print(bd.shape)
bd.head()

(15109, 7)


country_name  year  crude_birth_rate  crude_death_rate  net_migration  \
0     Slovenia  2036              7.59             14.65           0.23   
1     Slovenia  2022              7.52             12.28           0.32   
2     Slovenia  2023              7.43             12.41           0.31   
3     Slovenia  2024              7.35             12.55           0.31   
4     Slovenia  2025              7.30             12.68           0.30   

   rate_natural_increase  growth_rate  
0                 -0.706       -0.682  
1                 -0.476       -0.444  
2                 -0.498       -0.467  
3                 -0.520       -0.489  
4                 -0.538       -0.508

In [16]:
bd.tail(3)

country_name  year  crude_birth_rate  crude_death_rate  \
15106  Solomon Islands  1998             35.98              4.70   
15107  Solomon Islands  1997             36.44              4.79   
15108  Solomon Islands  1995             37.17              4.97   

       net_migration  rate_natural_increase  growth_rate  
15106          -2.45                  3.128        2.884  
15107          -2.31                  3.165        2.934  
15108          -2.01                  3.220        3.019

---

## Mortality Life Expentancy

In [17]:
mo = pd.read_csv(f'{path2}/mortality_life_expectancy.csv')
# I don't need all these columns, I'll take the average of each
mo.drop(columns={'country_code','infant_mortality_male','infant_mortality_female','life_expectancy_male',
                'life_expectancy_female','mortality_rate_under5_male','mortality_rate_under5_female',
                 'mortality_rate_1to4_male','mortality_rate_1to4_female'}, inplace=True)
print(mo.shape)
mo.head(10)

(15106, 6)


country_name  year  infant_mortality  life_expectancy  \
0     Slovenia  2036              3.39            80.90   
1     Slovenia  2022              3.76            79.11   
2     Slovenia  2023              3.73            79.26   
3     Slovenia  2024              3.70            79.40   
4     Slovenia  2025              3.67            79.55   
5     Slovenia  2026              3.64            79.68   
6     Slovenia  2027              3.61            79.81   
7     Slovenia  2028              3.58            79.94   
8     Slovenia  2029              3.56            80.08   
9     Slovenia  2030              3.53            80.21   

   mortality_rate_under5  mortality_rate_1to4  
0                   3.93                 0.54  
1                   4.43                 0.68  
2                   4.39                 0.67  
3                   4.35                 0.65  
4                   4.31                 0.64  
5                   4.26                 0.63  
6                   4.23                 0.63  
7                   4.19                 0.61  
8                   4.15                 0.60  
9                   4.12                 0.60

In [18]:
mo.tail(3)

country_name  year  infant_mortality  life_expectancy  \
15103  Solomon Islands  1998             27.45             70.7   
15104  Solomon Islands  1997             28.38             70.4   
15105  Solomon Islands  1995             30.33             69.8   

       mortality_rate_under5  mortality_rate_1to4  
15103                  34.71                 7.46  
15104                  35.97                 7.81  
15105                  38.64                 8.57

In [19]:
mo.country_name.value_counts()

Bhutan                 101
Djibouti               101
Nigeria                 98
Guinea                  96
Cambodia                89
                      ... 
Virgin Islands U.S.     51
South Sudan             43
Sudan                   43
Puerto Rico             41
United States           37
Name: country_name, Length: 228, dtype: int64

In [20]:
mo[mo.country_name == 'Bhutan']
print(mo[mo.country_name == 'Bhutan'].year.min())
print(mo[mo.country_name == 'Bhutan'].year.max())

1950
2050


In [21]:
mo[mo.country_name == 'United States']
print(mo[mo.country_name == 'United States'].year.min())
print(mo[mo.country_name == 'United States'].year.max())

2014
2050


In [22]:
len(mo[mo.year ==2014])

228

In [23]:
len(mo[mo.year == 2050])

228

In [24]:
mo[mo.year == 2014]

country_name  year  infant_mortality  life_expectancy  \
53                      Slovenia  2014              4.04            77.83   
116                       Latvia  2014              5.44            74.01   
178                     Mongolia  2014             23.15            68.98   
241                   Mauritania  2014             56.06            62.28   
276                      Liberia  2014             69.19            58.21   
...                          ...   ...               ...              ...   
14818                 Azerbaijan  2014             26.67            71.91   
14878  Saint Pierre and Miquelon  2014              6.95            80.26   
14933   Northern Mariana Islands  2014              5.50            77.64   
14959                       Chad  2014             90.30            49.44   
15049            Solomon Islands  2014             16.17            74.89   

       mortality_rate_under5  mortality_rate_1to4  
53                      4.83                 0.80  
116                     6.81                 1.37  
178                    27.78                 4.73  
241                    84.47                30.10  
276                   102.53                35.82  
...                      ...                  ...  
14818                  29.59                 3.00  
14878                   7.94                 0.99  
14933                   6.81                 1.32  
14959                 151.71                67.50  
15049                  19.71                 3.60  

[228 rows x 6 columns]

In [25]:
(mo.year.value_counts() == 228).sum()

37

---

## World Population

In [26]:
world = pd.read_csv('/Users/everettclark/Downloads/World Population Live Dataset.csv')
world.drop(columns='CCA3', inplace=True)
print(world.shape)
world.sort_values(by='Name').head()

(234, 14)


Name   2022   2020   2015   2010   2000   1990   1980   1970  \
35      Afghanistan  41129  38972  33753  28190  19543  10695  12487  10753   
137         Albania   2842   2867   2882   2913   3182   3295   2942   2325   
33          Algeria  44903  43452  39543  35856  30775  25518  18739  13796   
212  American Samoa     44     46     51     55     58     48     33     27   
202         Andorra     80     78     72     72     66     54     36     20   

     Area (km²)  Density (per km²)  GrowthRate World Population Percentage  \
35       652230            63.0587        1.03                       0.52%   
137       28748            98.8702        1.00                       0.04%   
33      2381741            18.8531        1.02                       0.56%   
212         199           222.4774        0.98                       0.00%   
202         468           170.5641        1.01                       0.00%   

     Rank  
35     36  
137   138  
33     34  
212   213  
202   203

In [27]:
world

Name     2022     2020     2015     2010     2000     1990  \
0               China  1425887  1424930  1393715  1348191  1264099  1153704   
1               India  1417173  1396387  1322867  1240614  1059634   870452   
2       United States   338290   335942   324608   311183   282399   248084   
3           Indonesia   275501   271858   259092   244016   214072   182160   
4            Pakistan   235825   227197   210969   194454   154370   115414   
..                ...      ...      ...      ...      ...      ...      ...   
229        Montserrat        4        5        5        5        5       11   
230  Falkland Islands        4        4        3        3        3        2   
231              Niue        2        2        2        2        2        3   
232           Tokelau        2        2        1        1        2        2   
233      Vatican City        1        1        1        1        1        1   

       1980    1970  Area (km²)  Density (per km²)  GrowthRate  \
0    982372  822534     9706961           146.8933        1.00   
1    696828  557501     3287590           431.0675        1.01   
2    223140  200328     9372610            36.0935        1.00   
3    148177  115228     1904569           144.6529        1.01   
4     80624   59291      881912           267.4018        1.02   
..      ...     ...         ...                ...         ...   
229      11      11         102            43.0392        0.99   
230       2       2       12173             0.3105        1.00   
231       4       5         260             7.4385        1.00   
232       2       2          12           155.9167        1.01   
233       1       1           1           510.0000        1.00   

    World Population Percentage  Rank  
0                        17.88%     1  
1                        17.77%     2  
2                         4.24%     3  
3                         3.45%     4  
4                         2.96%     5  
..                          ...   ...  
229                       0.00%   230  
230                       0.00%   231  
231                       0.00%   232  
232                       0.00%   233  
233                       0.00%   234  

[234 rows x 14 columns]

### Rank is by World Population Percentage
#### - Population is in the thousands
#### - Area is accurate (when I checked US sqkm)
#### - Density (Vatican is .44 sqkm and has a lower density than it should)

---
---

### Birth and Mortality merge

In [28]:
bam = pd.merge(bd,mo)
print(bam.shape)
bam.head()

(15106, 11)


country_name  year  crude_birth_rate  crude_death_rate  net_migration  \
0     Slovenia  2036              7.59             14.65           0.23   
1     Slovenia  2022              7.52             12.28           0.32   
2     Slovenia  2023              7.43             12.41           0.31   
3     Slovenia  2024              7.35             12.55           0.31   
4     Slovenia  2025              7.30             12.68           0.30   

   rate_natural_increase  growth_rate  infant_mortality  life_expectancy  \
0                 -0.706       -0.682              3.39            80.90   
1                 -0.476       -0.444              3.76            79.11   
2                 -0.498       -0.467              3.73            79.26   
3                 -0.520       -0.489              3.70            79.40   
4                 -0.538       -0.508              3.67            79.55   

   mortality_rate_under5  mortality_rate_1to4  
0                   3.93                 0.54  
1                   4.43                 0.68  
2                   4.39                 0.67  
3                   4.35                 0.65  
4                   4.31                 0.64

In [29]:
bd.head(2)

country_name  year  crude_birth_rate  crude_death_rate  net_migration  \
0     Slovenia  2036              7.59             14.65           0.23   
1     Slovenia  2022              7.52             12.28           0.32   

   rate_natural_increase  growth_rate  
0                 -0.706       -0.682  
1                 -0.476       -0.444

In [30]:
mo.head(2)

country_name  year  infant_mortality  life_expectancy  \
0     Slovenia  2036              3.39            80.90   
1     Slovenia  2022              3.76            79.11   

   mortality_rate_under5  mortality_rate_1to4  
0                   3.93                 0.54  
1                   4.43                 0.68

## Predict the top and bottom 5 populated countries?

### .csv list:

### Not needed:
##### - country_names_area.csv

##### - midyear_population_5yr_age_sex.csv

##### - midyear_population_age_country_code.csv

##### - age_specific_fertility_rates.csv
(15106, 13)

##### - midyear_population.csv (mp1)
(23028, 3)

#### Data Dictionary:

- Data dictionary.txt 

### Using:
##### - midyear_population_age_sex.csv
(30280, 105)
##### mortality_life_expectancy.csv (mo)
(15106, 15)
##### - birth_death_growth_rates.csv (bd)
(15109, 7)
#### mo+bd combined into bam

---

In [52]:
world.head(3)

Name     2022     2020     2015     2010     2000     1990  \
0          China  1425887  1424930  1393715  1348191  1264099  1153704   
1          India  1417173  1396387  1322867  1240614  1059634   870452   
2  United States   338290   335942   324608   311183   282399   248084   

     1980    1970  Area (km²)  Density (per km²)  GrowthRate  \
0  982372  822534     9706961           146.8933        1.00   
1  696828  557501     3287590           431.0675        1.01   
2  223140  200328     9372610            36.0935        1.00   

  World Population Percentage  Rank  
0                      17.88%     1  
1                      17.77%     2  
2                       4.24%     3

In [53]:
world.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 234 entries, 0 to 233
Data columns (total 14 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Name                         234 non-null    object 
 1   2022                         234 non-null    int64  
 2   2020                         234 non-null    int64  
 3   2015                         234 non-null    int64  
 4   2010                         234 non-null    int64  
 5   2000                         234 non-null    int64  
 6   1990                         234 non-null    int64  
 7   1980                         234 non-null    int64  
 8   1970                         234 non-null    int64  
 9   Area (km²)                   234 non-null    int64  
 10  Density (per km²)            234 non-null    float64
 11  GrowthRate                   234 non-null    float64
 12  World Population Percentage  234 non-null    object 
 13  Rank                

pd.pivot_table(mp1, values='midyear_population', index='country_name', columns='year')

In [56]:
pd.pivot_table(world, values=world['2022'], index=world.index, columns='Name')

KeyError: 1425887